workflows/{labels,reviewers}: fix concurrency groups for nested workflows

wolfgangwalther opened the pull request:
This [didn't work as intended](https://github.com/NixOS/nixpkgs/pull/413785#issuecomment-2970315246). When a workflow is run with `workflow_call`, it will have `github.workflow` set to the *parent* workflow. So the `caller` input that we passed, resulted in this concurrency key:

```
Eval-Eval-...
```

But that's bad, because the labels and reviewers workflows will cancel each other!

What we actually want is this:
- Label and Reviewers workflow should have different groups.
- Reviewers called via Eval and called directly via undraft should have *different* groups.

We can't use the default condition we use everywhere else, because `github.workflow` is the same for Label and Reviewers. Thus, we hardcode the workflow's name as well. This essentially means we have this as a key:

```
<name-of-running-workflow>-<name-of-triggering-workflow>-<name-of-event>-<name-of-head-branch>
```

This should do what we want.

## Things done

- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
workflows/{labels,reviewers}: fix concurrency groups for nested workflows

This didn't work as intended. When a workflow is run with
`workflow_call`, it will have `github.workflow` set to the *parent*
workflow. So the `caller` input that we passed, resulted in this
concurrency key:

```
Eval-Eval-...
```

But that's bad, because the labels and reviewers workflows will cancel
each other!

What we actually want is this:
- Label and Reviewers workflow should have different groups.
- Reviewers called via Eval and called directly via undraft should have
*different* groups.

We can't use the default condition we use everywhere else, because
`github.workflow` is the same for Label and Reviewers. Thus, we hardcode
the workflow's name as well. This essentially means we have this as a
key:

```
<name-of-running-workflow>-<name-of-triggering-workflow>-<name-of-event>-<name-of-head-branch>
```

This should do what we want.

Since workflows can be made reusable workflows later on, we add those
hardcoded names to *all* concurrency groups. This avoids copy&paste
errors later on.

MattSturgeon commented on .github/workflows/labels.yml:
```diff
@@ -7,19 +7,14 @@ name: "Label PR"
 
 concurrency:
-  group: ${{ inputs.caller }}-${{ github.workflow }}-${{ github.event_name }}-${{ github.head_ref || github.run_id }}
+  group: Labels-${{ github.workflow }}-${{ github.event_name }}-${{ github.head_ref || github.run_id }}
```
Do we want a comment clarifying that the seeminly repeated workflow name is actually intended. I could imagine a PR in the future removing `Lables-` string because it _should_ be the same as `${ github.workflow }}` :grin: 

My wording is horrible, but maybe you could come up with something better:

> Hardcode `Labels` because `github.workflow` is the calling workflow, which may or may not be a different workflow

On the other hand, this may not really be needed, as it is documented by git-blame.

wolfgangwalther committed:
workflows: prevent accidental cancelling of other PRs

This can happen when two PRs run at the same time, which come from
different forks, but have the same head branch name.

github.head_ref is suggested by GitHub's docs, but.. that's not really
useful for cases with forks.

MattSturgeon commented on .github/workflows/README.md:
```diff
@@ -18,3 +18,25 @@ Some architectural notes about key decisions and concepts in our workflows:
+
+## Concurrency Groups
+
+We use [GitHub's Concurrency Groups](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/control-the-concurrency-of-workflows-and-jobs) to cancel older jobs on force pushes to Pull Requests.
```
Do we only cancel on _force_ push or _any_ push?

MattSturgeon commented on .github/workflows/README.md:
```diff
@@ -18,3 +18,25 @@ Some architectural notes about key decisions and concepts in our workflows:
+
+- Because we want to run jobs for different events at same time, we add `github.event_name` to the key. This is the case for the `pull_request` which runs on changes to the workflow files to test the new files and the same workflow from the base branch run via `pull_request_event`.
+
+- We don't want workflows of different Pull Requests to cancel each other, so we include `github.event.pull_request.number`. The [GitHub docs](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/control-the-concurrency-of-workflows-and-jobs#example-using-a-fallback-value) show using `github.head_ref` for this purpose, but this doesn't work well with forks: Differ ...
```
> Different users could have the same head branch name in their forks and run CI for their PRs at the same time.

I assumed that `head_ref` would be `refs/pull/1234/head`. Is it actually the branch name? :eyes:

MattSturgeon commented on .github/workflows/README.md:
```diff
@@ -18,3 +18,25 @@ Some architectural notes about key decisions and concepts in our workflows:
+
+- We don't want workflows of different Pull Requests to cancel each other, so we include `github.event.pull_request.number`. The [GitHub docs](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/control-the-concurrency-of-workflows-and-jobs#example-using-a-fallback-value) show using `github.head_ref` for this purpose, but this doesn't work well with forks: Differ ...
+
+- Sometimes, there is no `pull_request.number`. That's the case for `push` or `workflow_run` events. In this case, we add a fallback of `github.run_id`, to make sure that those run to completion in any case.
```
This feels like something that could be situational. In some scenarios cancelling an older `push` job may actually be what we want.

IIUC adding `run_id` to a group effectively makes it guaranteed to be a singleton group; i.e. it'll never be cancelled by another run.

---

Regardless of the above feedback, the current wording could be tweaked:
```suggestion
- Sometimes, there is no `pull_request.number`. That's the case for `push` or `workflow_run` events. To ensure non-PR runs are never cancelled, we add a fallback of `github.run_id`. This is a unique value for each workflow run.
```

MattSturgeon commented on .github/workflows/README.md:
```diff
@@ -18,3 +18,25 @@ Some architectural notes about key decisions and concepts in our workflows:
+
+- Sometimes, there is no `pull_request.number`. That's the case for `push` or `workflow_run` events. In this case, we add a fallback of `github.run_id`, to make sure that those run to completion in any case.
+
+- Of course, we run multiple workflows at the same time, so we add `github.workflow` to the key. Otherwise workflows would cancel each other.
```
TIL; before your recent PRs I'd always assumed concurrency groups were intrinsically per-workflow, such that `workflow_a.yml` can't ever cancel `workflow_b.yml`. :see_no_evil:

philiptaron replied:
Thanks GitHub 🤦🏻

MattSturgeon commented on .github/workflows/README.md:
```diff
@@ -18,3 +18,25 @@ Some architectural notes about key decisions and concepts in our workflows:
+
+- Of course, we run multiple workflows at the same time, so we add `github.workflow` to the key. Otherwise workflows would cancel each other.
+
+- There is a special case for reusable workflows called via `workflow_call` - they will have `github.workflow` set to their parent workflow's name. Thus, they would cancel each other. That's why we additionally hardcode the name of the workflow as well.
```
:+1:

MattSturgeon commented on .github/workflows/README.md:
```diff
@@ -18,3 +18,25 @@ Some architectural notes about key decisions and concepts in our workflows:
+
+- There is a special case for reusable workflows called via `workflow_call` - they will have `github.workflow` set to their parent workflow's name. Thus, they would cancel each other. That's why we additionally hardcode the name of the workflow as well.
+
+This results in a key with the following semantics:
```
Should this be less prescriptive?

```suggestion
In practice, this means concurrency groups are usually defined with the following semantics:
```

philiptaron reviewed (approved):
Thank you for becoming an expert at this, @wolfgangwalther.
